### ISSUE TITLE
Task02: 回归

### ISSUE THREAD

**Issue by martinx** [martinx/leeml-notes#3]
Task02: 回归
Step 1: Model
Step 2: Goodness of Function
- 如何用图来表示
  - 线性
  - 非线性
Step 3: Best Function

Gradient Descent
只要 L 是可微分的，就可用 Gradient Descent

Step 3: Gradient Descent
  - 代码实现，一个变量w， 两个变量w,b
  - 求导
  - 学习率，是否变化，变化方案
  - Local optimal/global optimal
  - 图形化，可视化
  - compute 
  - convex 凸的；凸面的

Model Selection
- A more complex model yields lower error on training data
- A more complex model does not alwyas lead to better performance on testing data
This is Overfitting

Why Overfitting

Model  不是越复杂越好，选择合适的 Model

Back to step1: Model
Hidden factors
会发现曲线按类别聚合在一起，还需要考虑更多的 feature，如分类
那么根据分类重新设计 Model（Step 1）

Back to step 2: Regularization
重新定义 L Loss Fuction

Gradient Descent
Overfitting
Regularization

Overestimate
Underestimate

微分
偏微分

多个特征，gradient

等高线
问题：
1: 当前最优 Stuck at local minima
2: 等于 0 Stuck at saddle point
3: 趋势于 0 Very slow at the plateau

偏微分计算方法

如何验证训练好的模型好坏
使用训练集和测试集的平均误差来验证模型的好坏

模型调整
当发现 average error  偏差比较大时，可以考虑

- 选择更复杂的模型，如更高次模型
什么是线性模型
一定是直线吗？

过拟合
Overfitting
在模型上，我们再进行进一步优化，使用更高此模型，会发现到后面训练集表现的很优秀，但是到测试集上的效果变的更差了，这就是模型在训练集上过拟合的问题。

步骤优化
发现数据偏差交大，研究发现其他的隐藏的特征

Step 1 优化：
将多个线性模型合并到一个线性模型中

Step 2优化：如果希望模型更强大表现更好（更多参数，更多 input）

更多特征，更多input，数据量没有明显增加，仍旧导致overfitting

Step 3 优化：加入正则化
更多特征，但是 权重 w 可能会使某些特征权重过高，仍然导致 overfitting，所以加入正则化

- w 越小，表示 function 较平滑的， function输出值与输入值相差不大
- 在很多应用场景中，并不是 w 越小模型越平滑越好，但是经验值告诉我们 w 越小大部分情况下都是好的。
- b 的值接近于0 ，对曲线平滑是没有影响

CODE
numpy.meshgrid
生成网格点坐标矩阵

A,B,C,D,E,F是6个网格点，坐标如图，如何用矩阵形式（坐标矩阵）来批量描述这些点的坐标呢？
答案如下：

原文链接：https://blog.csdn.net/lllxxq141592654/article/details/81532855

```
import numpy as np
import matplotlib.pyplot as plt

x_data = [338., 333., 328., 207., 226., 25., 179., 60., 208., 606.]
y_data = [640., 633., 619., 393., 428., 27., 193., 66., 226., 1591.]

x = np.arange(-200, -100, 1) #bias
y = np.arange(-5, 5, 0.1) #weight

Z = np.zeros((len(x), len(y)))
X, Y = np.meshgrid(x, y)

for i in range(len(x)):
    for j in range(len(y)):
        b = x[i]
        w = y[j]
        Z[j][i] = 0
        
        for n in range(len(x_data)):
            Z[j][i] += (y_data[n] - b - w * x_data[n]) ** 2
        Z[j][i] /= len(x_data)


b = -120
w = -4
lr = 0.0000001
iteration = 100_000

b_history = [b]
w_history = [w]

for i in range(iteration):
    
    b_grad = 0.0
    w_grad = 0.0
    
    for n in range(len(x_data)):
        b_grad = b_grad - 2.0 * (y_data[n] - b - w * x_data[n]) * -1.0
        w_grad = w_grad - 2.0 * (y_data[n] - b - w * x_data[n]) * x_data[n]
        
        b -= lr * b_grad
        w -= lr * w_grad
        
        b_history.append(b)
        w_history.append(w)
        
plt.contourf(x, y, Z, 50, alpha=0.5,  cmap=plt.get_cmap('jet'))
plt.plot([-188.4], [2.67], 'x', ms=12, markeredgewidth=3, color='orange')
plt.plot(b_history, w_history, 'o-', ms=3, lw=1.5, color='black')
plt.xlim(-200, -100)
plt.ylim(-5, 5)
plt.xlabel(r'$b$', fontsize=16)
plt.ylabel(r'$w$', fontsize=16)
plt.show()

```
